caddyconfig/httpcaddyfile/serveroptions.go

```diff
@@ -0,0 +1,247 @@
+// Copyright 2015 Matthew Holt and The Caddy Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package httpcaddyfile
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/caddyserver/caddy/v2"
+	"github.com/caddyserver/caddy/v2/caddyconfig"
+	"github.com/caddyserver/caddy/v2/caddyconfig/caddyfile"
+	"github.com/caddyserver/caddy/v2/modules/caddyhttp"
+	"github.com/dustin/go-humanize"
+)
+
+// serverOptions collects server config overrides parsed from Caddyfile global options
+type serverOptions struct {
+	// If set, will only apply to servers that contain a listener
+	// address that matches. If empty, will apply to all servers.
+	// Only the most specific one will apply.
+	ListenerMatcher string
```

Comment:
This must be a listener _address_ right? Should we rename this to `ListenerAddress`?

How do you define specificity?

Reply:
I guess `ListenerAddressMatcher`? Specifically it's the string that'll be used to match against the list of the server's listener addresses. I think it makes sense to have `Matcher` in the name for that reason.

Reply:
Crucially, this field matches one and only 1 listener, which must be an exact match, right? So `:8080` won't match all listeners on port 8080?

Reply:
No, `:8080` won't match `127.0.0.1:8080` (for now). We could set up `*:8080` or something if you feel it's worth it, but I think that's uncommon in the Caddyfile, because `bind` is not a commonly used feature I'd say.

Reply:
But yeah - specificity is length of the matcher, sorts them by listener matcher (empty, i.e. "any", is always last)

Reply:
Exact matching is fine, but it's hardly "matching" at that point -- it's literally just the address of the listener. Matching implies some smart or cascading logic IMO.

If specificity is length, it's hard to say that `:8080` is really less specific than `:80` - the two are precisely equally specific. I think we should lose the notions of specificity and matching here, and just say that it's the address of the server listener which we're applying this to.

Reply:
The reason I mentioned specificity is because `""` is decidedly less specific than `:80` (and will apply to "everything else").

But yeah I guess I can reword it.